PROJECTION expression

n9iels opened the pull request:
This PR adds support for the `PROJECTION`, an additional expression that can be used together with the `CREATE` expression.

The  expression consists out of the properties `context` and `fact`. The `context` property is an array to allow the usage of nested `CREATE` expressions.
```json
{
    "expression": "PROJECTION",
    "context": [
        "[aanvraag]"
    ],
    "fact": "[bedrag]"
}
```

Model example:
```json
{
    "acts": [
        {
            "act": "<<subsidie aanvragen>>",
            "actor": "[burger]",
            "recipient": "[ambtenaar]",
            "object": "[verzoek]",
            "preconditions": {
                "expression": "AND",
                "operands": [
                    "[bedrag]"
                ]
            },
            "create": [
                "[aanvraag]"
            ]
        },
        {
            "act": "<<subsidie aanvraag toekennen>>",
            "actor": "[ambtenaar]",
            "object": "[aanvraag]",
            "preconditions": {
                "expression": "LESS_THEN",
                "operands": [
                    {
                        "expression": "PROJECTION",
                        "context": [
                            "[aanvraag]"
                        ],
                        "fact": "[bedrag]"
                    },
                    {
                        "expression": "LITERAL",
                        "operand": 500
                    }
                ]
            },
            "recipient": "[burger]"
        }
    ],
    "facts": [
        {
            "fact": "[aanvraag]",
            "function": {
                "expression": "CREATE",
                "operands": [
                    "[bedrag]"
                ]
            }
        }
    ],
    "duties": []
}
```

pimotte commented on test/index.spec.js:
```diff
@@ -1941,4 +1888,450 @@ describe('discipl-law-reg', () => {
+            'recipient': '[ambtenaar]',
+            'object': '[verzoek]',
+            'preconditions': {
+              'expression': 'AND',
```
Any particular reason this has the "AND" operator wrapping the operand?

n9iels replied:
No :) Usage of `AND` is indeed redundant in this case

n9iels commented:
@pimotte do these new expressions need to be implemented in https://github.com/discipl/flinteditor-vscode too?

pimotte commented on test/flint-example-lerarenbeurs.json:
```diff
@@ -959,7 +959,10 @@
-      "function": { "expression": "CREATE" },
+      "function": {
+        "expression": "CREATE",
+        "operands": []
```
Do you think we should make operands optional? I think the trade-off here is that we don't have this optionality in other places, but it avoids introducing a breaking change.

n9iels replied:
I agree. Also, the expression it useful without any operands. So it makes sense to me

pimotte commented:
> @pimotte do these new expressions need to be implemented in https://github.com/discipl/flinteditor-vscode too?

I'm tempted to leave it and instead work with @blackstardlb to get it supported in flinteditor-mps instead.

n9iels commented:
> Document the new operations

I have added some documentation to https://github.com/discipl/discipl-law-reg/pull/116 and made it ready for review.

> Test for OR in CREATE expression

I have added a test where a `OR` expression is used withing `CREATE` expressions.

pimotte reviewed (approved):
Reviewed this before, content is good:D

Style/structure improvements across the board I'm happy with too, well done! Props to both of you, @n9iels and @blackstardlb
